crates/sui-types/src/messages.rs

```diff
@@ -1062,10 +1090,149 @@ impl PartialEq for SignedTransaction {
+///   TransactionEnvelope<T> into a VerifiedTransactionEnvelope<T>
+/// - From<TrustedTransactionEnvelope<T>> - used when reading pre-verified messages from the
+///   database.
+/// - VerifiedTransactionEnvelope::new_unchecked() - use very carefully, and very sparingly, so
+///   that all uses can be audited easily.
+#[derive(Clone)]
+pub struct VerifiedTransactionEnvelope<T>(TrustedTransactionEnvelope<T>, NoSer);
+
+// Never remove this assert!
+static_assertions::assert_not_impl_any!(VerifiedTransactionEnvelope: Serialize, DeserializeOwned);
+
+impl<T> Debug for VerifiedTransactionEnvelope<T>
+where
+    T: Debug,
+{
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{:?}", self.0 .0)
+    }
+}
+
+impl<T> VerifiedTransactionEnvelope<T> {
+    fn new_from_verified(inner: TransactionEnvelope<T>) -> Self {
+        Self(TrustedTransactionEnvelope(inner), NoSer)
+    }
+
+    /// There are some situations (e.g. fragment verification) where its very awkward and/or
+    /// inefficient to obtain verified certificates from calling CertifiedTransaction::verify()
+    /// Use this carefully.
+    pub fn new_unchecked(inner: TransactionEnvelope<T>) -> Self {
+        Self(TrustedTransactionEnvelope(inner), NoSer)
+    }
+
+    pub fn into_inner(self) -> TransactionEnvelope<T> {
+        self.0 .0
+    }
+
+    /// Use this when you need to serialize a verified envelope.
+    /// This should generally only be used for database writes.
+    /// ***never use over the network!***
+    pub fn serializable_ref(&self) -> &TrustedTransactionEnvelope<T> {
+        &self.0
+    }
+
+    /// Use this when you need to serialize a verified envelope.
+    /// This should generally only be used for database writes.
+    /// ***never use over the network!***
+    pub fn serializable(self) -> TrustedTransactionEnvelope<T> {
+        self.0
+    }
+}
+
+/// After deserialization, a TrustedTransactionEnvelope can be turned back into a
+/// VerifiedTransactionEnvelope.
+impl<T> From<TrustedTransactionEnvelope<T>> for VerifiedTransactionEnvelope<T> {
+    fn from(e: TrustedTransactionEnvelope<T>) -> Self {
+        Self::new_unchecked(e.0)
+    }
+}
+
+impl<T> Deref for VerifiedTransactionEnvelope<T> {
+    type Target = TransactionEnvelope<T>;
+    fn deref(&self) -> &Self::Target {
+        &self.0 .0
+    }
+}
+
+impl<T> DerefMut for VerifiedTransactionEnvelope<T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0 .0
+    }
+}
+
+impl<T> From<VerifiedTransactionEnvelope<T>> for TransactionEnvelope<T> {
+    fn from(v: VerifiedTransactionEnvelope<T>) -> Self {
+        v.0 .0
+    }
+}
+
+impl<T> PartialEq for VerifiedTransactionEnvelope<T>
+where
+    TransactionEnvelope<T>: PartialEq,
+{
+    fn eq(&self, other: &Self) -> bool {
+        self.0 .0 == other.0 .0
+    }
+}
+
+impl<T> Eq for VerifiedTransactionEnvelope<T> where TransactionEnvelope<T>: Eq {}
```

Comment:
Are these impls needed? Also I think that if we do want them we'd need to have the bounds `T: PartialEq` and `T: Eq` as well right?

Reply:
You're missing those bounds on the Eq impl